### PR TITLE
fix(codegen): MISRA Rule 8.5 - remove extern from const definitions in C mode (#852)

### LINKED ISSUE #852: MISRA C:2012 Rule 8.5: External declaration in one file only (6 violations)
## MISRA Rule

**Rule 8.5 (Mandatory)**: An external object or function shall be declared once in one and only one file.

## Violation Count

6 violations across test files.

## Description

External declarations should appear in exactly one header file.

## Example Violation

```c
// file1.c
extern int global_var;  // Violation if also declared elsewhere

// Should be in one header only:
// globals.h: extern int global_var;
```

## Fix Strategy

Review header generation to ensure single declaration point for externals.

## Related

- Part of #839 breakdown
- MISRA C:2012 Mandatory rule

### PR DESCRIPTION
**Pull request by jlaustill**
## Summary

- Fix MISRA C:2012 Rule 8.5 violations by removing `extern` from const definitions in C mode
- Maintain `extern` in C++ mode for external linkage (Issue #525 compatibility)
- Update CLAUDE.md with MISRA validation and codegen tips

### Code Changes
- `VariableModifierBuilder.ts`: Add `hasInitializer` and `cppMode` parameters to conditionally add `extern`
- `VariableDeclHelper.ts`: Pass `hasInitializer` and `cppMode` to the modifier builder

### Test Changes
- Added unit tests for both C and C++ mode behavior
- Updated all 972 test snapshots with correct C behavior

### Documentation
- Updated `docs/misra-compliance.md` with Issue #852 reference
- Added MISRA validation tips to `CLAUDE.md`

## Root Cause

`VariableModifierBuilder.build()` was unconditionally adding `extern` to all top-level const variables. When the .c file includes its header (which declares `extern const ...`), the definition also had `extern`, creating duplicate declarations.

## Solution

- **C mode**: `extern` only for declarations (no initializer), NOT definitions
- **C++ mode**: Always `extern` (needed for external linkage since C++ const has internal linkage by default)

## Results

| Metric | Before | After |
|--------|--------|-------|
| Rule 8.5 violations | 31+ | 0 |
| Unit tests | 5413 | 5415 |
| Integration tests | 972 | 972 |

## Test plan

- [x] All 5415 unit tests pass
- [x] All 972 integration tests pass
- [x] Zero MISRA Rule 8.5 violations
- [x] C++ cross-file const linkage still works (Issue #525)

Closes #852

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/transpiler/output/codegen/helpers/VariableModifierBuilder.ts [hunk 1]
```diff
@@ -46,20 +46,35 @@ class VariableModifierBuilder {
...
+   * @param hasInitializer - Whether the variable has an initializer (affects extern in C mode)
+   * @param cppMode - Whether we're generating C++ code (affects extern behavior)
    * @returns Modifier strings ready for use in generated code
    * @throws Error if both atomic and volatile are specified
    */
   static build(
     ctx: IModifierContext,
     inFunctionBody: boolean,
+    hasInitializer: boolean = false,
+    cppMode: boolean = false,
```
Comment (jlaustill) [inline src/transpiler/output/codegen/helpers/VariableModifierBuilder.ts]: **[moderate]** testing: Consider adding a test case for the edge case where `hasConst && !inFunctionBody && !cppMode && !hasInitializer` — a const declaration without initializer in C mode (e.g., `const u32 EXTERNAL_VALUE;` without assignment).

The existing test "returns extern for const at file scope (declaration without initializer, C mode)" covers this, but it might be worth having an integration test that demonstrates the header+implementation split pattern that MISRA Rule 8.5 expects.
